[SPMD] Upstream partial_lower

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99076
* #99075
* __->__ #99069

Several ops cannot be lowered to the Inductor. This PR copies the internal implementation of partial_lower (credit to @yifuwang ) to torch.distributed._spmd to unblock the OSS usage. The internal version will be kept until it is mature and will replace this version.

Differential Revision: [D44970278](https://our.internmc.facebook.com/intern/diff/D44970278/)

fegin committed:
[SPMD] Upstream partial_lower

Several ops cannot be lowered to the Inductor. This PR copies the internal implementation of partial_lower (credit to @yifu) to torch.distributed._spmd to unblock the OSS usage. The internal version will be kept until it is mature and will replace this version.

Differential Revision: [D44970278](https://our.internmc.facebook.com/intern/diff/D44970278/)

[ghstack-poisoned]

mrshenli reviewed (approved):
Do we still need `lower_to_inductor` or should we remove it?

fegin commented:
> Do we still need `lower_to_inductor` or should we remove it?

Will remove the unused code in a separate PR.

lessw2020 commented on torch/distributed/_spmd/gm_transformation.py:
```diff
@@ -17,6 +17,7 @@
 )
 from torch.distributed._spmd.graph_utils import dump_graphs_to_files, OP
 from torch.distributed._spmd.iter_graph_module import IterGraphModule
+from torch.distributed._spmd.partial_lower import partial_lower
```
maybe worth a comment that this is expected to be manually copied over? (to avoid confusion)

fegin replied:
I added the comment in `partial_lower.py`

lessw2020 commented on test/distributed/_spmd/test_transformation.py:
```diff
@@ -79,7 +79,9 @@ def _init(self, batch_size, layers, dim, foreach: bool = True, fused: bool = Fal
         return model, optim, ddp_model, ddp_optim
 
-    def _test_tran_step_with_ddp(self, train_step, num_iters, batch_size, layers, dim):
+    def _test_tran_step_with_ddp(
```
minor nit - rename to _train_step instead of _tran_step.

lessw2020 reviewed (approved):
looks great - thanks for doing this.
added two minor nits mostly for future user clarity.

fegin committed:
Update on "[SPMD] Upstream partial_lower"


Several ops cannot be lowered to the Inductor. This PR copies the internal implementation of partial_lower (credit to yifuwang ) to torch.distributed._spmd to unblock the OSS usage. The internal version will be kept until it is mature and will replace this version.

Differential Revision: [D44970278](https://our.internmc.facebook.com/intern/diff/D44970278/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Upstream partial_lower"


Several ops cannot be lowered to the Inductor. This PR copies the internal implementation of partial_lower (credit to yifuwang ) to torch.distributed._spmd to unblock the OSS usage. The internal version will be kept until it is mature and will replace this version.

Differential Revision: [D44970278](https://our.internmc.facebook.com/intern/diff/D44970278/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Upstream partial_lower"


Several ops cannot be lowered to the Inductor. This PR copies the internal implementation of partial_lower (credit to yifuwang ) to torch.distributed._spmd to unblock the OSS usage. The internal version will be kept until it is mature and will replace this version.

Differential Revision: [D44970278](https://our.internmc.facebook.com/intern/diff/D44970278/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Upstream partial_lower"


Several ops cannot be lowered to the Inductor. This PR copies the internal implementation of partial_lower (credit to yifuwang ) to torch.distributed._spmd to unblock the OSS usage. The internal version will be kept until it is mature and will replace this version.

Differential Revision: [D44970278](https://our.internmc.facebook.com/intern/diff/D44970278/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Upstream partial_lower"


Several ops cannot be lowered to the Inductor. This PR copies the internal implementation of partial_lower (credit to yifuwang ) to torch.distributed._spmd to unblock the OSS usage. The internal version will be kept until it is mature and will replace this version.

Differential Revision: [D44970278](https://our.internmc.facebook.com/intern/diff/D44970278/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
